pkg/capi/gcp.go

```diff
@@ -0,0 +1,306 @@
+				clusterName,
+				mapiMachineSpec.Zone,
+				1,
+				corev1.ObjectReference{
+					Kind:       "GCPMachineTemplate",
+					APIVersion: infraAPIVersion,
+					Name:       gcpMachineTemplate.Name,
+				},
+			))
+			Expect(err).ToNot(HaveOccurred(), "Failed to create CAPI machineset with Shielded VM config")
+
+			framework.WaitForCAPIMachinesRunning(framework.GetContext(), cl, machineSet.Name)
+
+			By("Verifying the Shielded VM configuration on the created GCP MachineTemplate")
+			createdTemplate := &gcpv1.GCPMachineTemplate{}
+			Expect(cl.Get(ctx, client.ObjectKey{
+				Namespace: framework.ClusterAPINamespace,
+				Name:      gcpMachineTemplate.Name,
+			}, createdTemplate)).To(Succeed())
+			Expect(createdTemplate.Spec.Template.Spec.ShieldedInstanceConfig).ToNot(BeNil())
+			Expect(fmt.Sprintf("%v", createdTemplate.Spec.Template.Spec.ShieldedInstanceConfig.SecureBoot)).To(Equal(fmt.Sprintf("%v", enableSecureBoot)))
+			Expect(fmt.Sprintf("%v", createdTemplate.Spec.Template.Spec.ShieldedInstanceConfig.VirtualizedTrustedPlatformModule)).To(Equal(fmt.Sprintf("%v", enableVtpm)))
+			Expect(fmt.Sprintf("%v", createdTemplate.Spec.Template.Spec.ShieldedInstanceConfig.IntegrityMonitoring)).To(Equal(fmt.Sprintf("%v", enableIntegrityMonitoring)))
+		},
+		Entry("all Shielded VM options enabled", gcpv1.SecureBootPolicyEnabled, gcpv1.VirtualizedTrustedPlatformModulePolicyEnabled, gcpv1.IntegrityMonitoringPolicyEnabled),
+		Entry("only SecureBoot enabled", gcpv1.SecureBootPolicyEnabled, gcpv1.VirtualizedTrustedPlatformModulePolicyDisabled, gcpv1.IntegrityMonitoringPolicyDisabled),
+		/*Below configs doesn't make difference due to defaulting conditions of shielded VMs
+		Entry("only Vtpm enabled", gcpv1.SecureBootPolicyDisabled, gcpv1.VirtualizedTrustedPlatformModulePolicyEnabled, gcpv1.IntegrityMonitoringPolicyDisabled),
+		Entry("only IntegrityMonitoring enabled", gcpv1.SecureBootPolicyDisabled, gcpv1.VirtualizedTrustedPlatformModulePolicyDisabled, gcpv1.IntegrityMonitoringPolicyEnabled),
+		Entry("SecureBoot and Vtpm enabled", gcpv1.SecureBootPolicyEnabled, mapiv1.VirtualizedTrustedPlatformModulePolicyEnabled, gcpv1.IntegrityMonitoringPolicyDisabled),
+		Entry("SecureBoot and IntegrityMonitoring enabled", gcpv1.SecureBootPolicyEnabled, gcpv1.VirtualizedTrustedPlatformModulePolicyDisabled, gcpv1.IntegrityMonitoringPolicyEnabled),
+		Entry("all Shielded VM options disabled", gcpv1.SecureBootPolicyDisabled, gcpv1.VirtualizedTrustedPlatformModulePolicyDisabled, gcpv1.IntegrityMonitoringPolicyDisabled),
+		*/
+	)
+	// Define constants as variable for this case to pass values properly
+	confidentialComputeEnabled := gcpv1.ConfidentialComputePolicyEnabled
+	confidentialComputeDisabled := gcpv1.ConfidentialComputePolicyDisabled
+	DescribeTable("should configure Confidential VM correctly", framework.LabelCAPI, framework.LabelDisruptive,
+		func(confidentialCompute *gcpv1.ConfidentialComputePolicy) {
+			mapiProviderSpec := getGCPMAPIProviderSpec(cl)
```

Comment:
What is this providerSpec used for? We are setting the onhostmaintenance stuff, but we don't actually apply it to the cluster?

Reply:
We apply it when ConfidentialCompute is enabled.

Reply:
I don't see that? You're creating a GCP Machine template, and a GCP MachineSet, I don't see you using a MAPI providerSpec anywhere as you aren't creating MAPI machines?

Reply:
yes Joel you are correct  , but populating values for gcpmachinetemplate eg- zone with machineproviderspec. 
so I was making changes to confidentialcompute fields and using the rest of the values to make the template similar to machineproviderspec from mapi , this was done on all tests.

Reply:
Right I see what's going on here, but then why do you modify the mapiProviderSpec after line 150?

Should line 150 come after the switch that updates the MAPI provider spec?

Reply:
so does making below change  is same instead of moving 150 down , I can do below? You feel, it might be redundant code at this moment ? I am not sure , how I reached there :thinking: 
```
			switch confidentialCompute {
			case gcpv1.ConfidentialComputePolicyEnabled:
				gcpMachineTemplate.Spec.Template.Spec.OnHostMaintenance = ptr.To(gcpv1.HostMaintenancePolicy("onHostMaintenance"))
			case gcpv1.ConfidentialComputePolicyDisabled:
				gcpMachineTemplate.Spec.Template.Spec.OnHostMaintenance = ptr.To(gcpv1.HostMaintenancePolicy("Migrate"))
			}

```

Reply:
Yeah, I think the switch is redundant. You convert the MAPI provider spec to the GCP Template spec on 150, after that you don't re-use the mapi provider spec so any further changes aren't being considered. I would have expected all that setup to come ahead of the conversion

Reply:
If I try to do that afterwards , it will result in panic due to gcpmachinetemplate being empty ?

Reply:
Yes, my expectation was to move the setting of any of the gcpMachineTemplate into the conversion function `createGCPMachineTemplate`, I see in most tests you're modifying the `gcpMachineTemplate` after you've got the result of that function.

Either you need to modify the mapi, or the capi one, if you modify the mapi one, then do that before you pass it into `createGCPMachineTemplate` and ensure that `createGCPMachineTemplate` handles all the fields you're setting.

Or remove modifications to `mapiProviderSpec` after you call `createGCPMachineTemplate`, those modifications are not being read apart from in the lines directly after them, in which case you may as well cut out the middle step no?